File: src/ipc/export.ts
import { invoke } from "@tauri-apps/api/core";
import { save } from "@tauri-apps/plugin-dialog";

export type ExportFormat = "png" | "jpeg" | "webp";

export interface PrivacyRegionExport {
  region_type: "blur" | "pixelate";
  x: number;
  y: number;
  width: number;
  height: number;
  intensity: number;
}

export interface ExportOptions {
  format: ExportFormat;
  quality: number;
  scale: number;
  privacyRegions?: PrivacyRegionExport[];
}

const FORMAT_FILTERS: Record<ExportFormat, { name: string; extensions: string[] }[]> = {
  png: [{ name: "PNG Image", extensions: ["png"] }],
  jpeg: [{ name: "JPEG Image", extensions: ["jpg", "jpeg"] }],
  webp: [{ name: "WebP Image", extensions: ["webp"] }],
};

/**
 * Show a native save dialog, then export canvas pixel data via Rust.
 */
export async function exportCanvas(
  imageData: Uint8Array,
  width: number,
  height: number,
  options: ExportOptions,
): Promise<string | null> {
  const path = await save({
    defaultPath: `screenshot.${options.format}`,
    filters: FORMAT_FILTERS[options.format],
  });

  if (!path) return null;

  return invoke<string>("export_image", {
    imageData: Array.from(imageData),
    width,
    height,
    outputPath: path,
    format: options.format,
    quality: options.quality,
    scale: options.scale,
    privacyRegions: options.privacyRegions ?? null,
  });
}

File: src-tauri/src/commands/export.rs
use crate::processing;
use image::{ImageBuffer, Rgba};
use std::path::PathBuf;

/// Export raw RGBA image data to a file in the specified format.
/// The save dialog is handled on the frontend via @tauri-apps/plugin-dialog.
#[tauri::command]
pub async fn export_image(
    image_data: Vec<u8>,
    width: u32,
    height: u32,
    output_path: String,
    format: String,
    quality: u32,
    scale: u32,
) -> Result<String, String> {
    let img: ImageBuffer<Rgba<u8>, Vec<u8>> =
        ImageBuffer::from_raw(width, height, image_data)
            .ok_or_else(|| "Invalid image data dimensions".to_string())?;

    let (final_img, _final_w, _final_h) = processing::scale_image(img, scale);

    let path = PathBuf::from(&output_path);
    processing::encode_to_file(&final_img, &path, &format, quality)?;

    Ok(output_path)
}
